File: applications/node/mcp.mjs
import {spawn} from 'node:child_process';
import {readFile} from 'node:fs/promises';
import path from 'node:path';
import {randomUUID} from 'node:crypto';
import {strictJSON, sha256, fail, Failure, object, deepEqual, VERSION} from './base.mjs';

export const PROTOCOL='2025-11-25';
const LIMIT=4*1024*1024;
const META='engineering-process-provider';

export class MCPClient {
  constructor(transport,{directory=process.cwd(),timeout=15000}={}) {
    this.transport=transport;this.directory=directory;this.timeout=timeout;this.pending=new Map();this.closed=false;this.session=null;
  }
  async start() {
    if(this.transport.kind==='http')return;
    if(this.transport.kind!=='stdio')fail('EPX_TRANSPORT_UNSUPPORTED','unsupported','BIND-002','Unsupported MCP transport.');
    this.child=spawn(this.transport.command,this.transport.arguments??[],{cwd:this.directory,shell:false,stdio:['pipe','pipe','pipe'],env:process.env});
    this.child.stderr.resume(); // Provider stderr may contain secrets; do not put it in public reports.
    let buffer='';
    this.child.stdout.setEncoding('utf8');
    this.child.stdout.on('data',chunk=>{
      buffer+=chunk;
      if(Buffer.byteLength(buffer)>LIMIT){this.abort(new Error('MCP message exceeds bounded size.'));return;}
      for(;;){const end=buffer.indexOf('\n');if(end<0)break;const line=buffer.slice(0,end);buffer=buffer.slice(end+1);if(!line.trim())continue;
        let message;try{message=strictJSON(line);}catch(error){this.abort(error);return;}
        if(!object(message)||message.jsonrpc!=='2.0'){this.abort(new Error('Invalid JSON-RPC response envelope.'));return;}
        if(!Object.hasOwn(message,'id'))continue;
        const pending=this.pending.get(message.id);if(!pending)continue;
        this.pending.delete(message.id);clearTimeout(pending.timer);pending.resolve(message);
      }
    });
    this.child.on('error',error=>this.abort(error));
    this.child.on('exit',()=>{this.closed=true;this.abort(new Error('MCP transport closed before a response.'));});
  }
  abort(error) {
    for(const {reject,timer} of this.pending.values()){clearTimeout(timer);reject(error);}this.pending.clear();
  }
  async exchange(message) {
    if(this.transport.kind==='http') {
      const headers={'Content-Type':'application/json','Accept':'application/json, text/event-stream','MCP-Protocol-Version':PROTOCOL};
      if(this.session)headers['Mcp-Session-Id']=this.session;
      if(this.transport.tokenEnv){const value=process.env[this.transport.tokenEnv];if(!value)throw new Error('Configured credential environment variable is unavailable.');headers.Authorization=`Bearer ${value}`;}
      const response=await fetch(this.transport.url,{method:'POST',headers,body:JSON.stringify(message),signal:AbortSignal.timeout(this.timeout),redirect:'error'});
      if(response.headers.has('mcp-session-id'))this.session=response.headers.get('mcp-session-id');
      if(!response.ok)throw new Failure('EPX_HTTP_FAILED','unavailable','BIND-002','MCP HTTP transport rejected the request.',{cause:{status:response.status}});
      if(!Object.hasOwn(message,'id')){await response.arrayBuffer();return null;}
      if(!response.headers.get('content-type')?.includes('application/json'))fail('EPX_HTTP_STREAM_UNSUPPORTED','unsupported','BIND-002','This adapter supports MCP HTTP JSON responses; negotiated SSE is unsupported.');
      const text=await response.text();if(Buffer.byteLength(text)>LIMIT)throw new Error('MCP response exceeds bounded size.');
      return strictJSON(text);
    }
    if(this.closed||!this.child?.stdin.writable)throw new Error('MCP transport is unavailable.');
    if(!Object.hasOwn(message,'id')){this.child.stdin.write(JSON.stringify(message)+'\n');return null;}
    return await new Promise((resolve,reject)=>{
      const timer=setTimeout(()=>{this.pending.delete(message.id);reject(new Error('MCP response timed out.'));},this.timeout);
      this.pending.set(message.id,{resolve,reject,timer});
      this.child.stdin.write(JSON.stringify(message)+'\n',error=>{if(error){const item=this.pending.get(message.id);if(item){clearTimeout(item.timer);this.pending.delete(message.id);item.reject(error);}}});
    });
  }
  async request(method,params={}) {
    const request={jsonrpc:'2.0',id:randomUUID(),method,params};
    this.lastRequest=request;
    const response=await this.exchange(request);
    this.lastResponse=response;
    if(!object(response)||response.jsonrpc!=='2.0'||response.id!==request.id||Object.hasOwn(response,'result')===Object.hasOwn(response,'error'))
      fail('EPX_RPC_INVALID','execution-failed','OUT-003','MCP response is not the matching JSON-RPC result/error envelope.');
    if(response.error)fail('EPX_RPC_ERROR','execution-failed','OUT-003','MCP returned a JSON-RPC error.',{cause:response.error});
    return response.result;
  }
  async initialize() {
    await this.start();
    const result=await this.request('initialize',{protocolVersion:PROTOCOL,capabilities:{},clientInfo:{name:'epx-node',version:VERSION}});
    if(result?.protocolVersion!==PROTOCOL)fail('EPX_PROTOCOL_MISMATCH','unavailable','BIND-003','MCP protocol negotiation does not match the declared revision.',{expected:PROTOCOL,observed:result?.protocolVersion});
    if(!object(result.capabilities?.tools))fail('EPX_TOOLS_UNAVAILABLE','unavailable','BIND-003','Provider does not advertise MCP tools capability.');
    await this.exchange({jsonrpc:'2.0',method:'notifications/initialized'});
    return result;
  }
  async close() {
    this.abort(new Error('MCP session closed.'));
    this.child?.stdin.end();this.child?.kill();this.closed=true;
    // HTTP DELETE is session cleanup only; it never retries engineering operations.
    if(this.transport.kind==='http'&&this.session) {
      const headers={'Mcp-Session-Id':this.session,'MCP-Protocol-Version':PROTOCOL};
      if(this.transport.tokenEnv&&process.env[this.transport.tokenEnv])headers.Authorization=`Bearer ${process.env[this.transport.tokenEnv]}`;
      try{await fetch(this.transport.url,{method:'DELETE',headers,signal:AbortSignal.timeout(1000),redirect:'error'});}catch{/* Unknown cleanup cannot authorize replay. */}
    }
  }
}

function providerMetadata(response,intended,dependency) {
  const observed=response?._meta?.[META];
  const expected={implementation:intended.implementation.id,revision:intended.implementation.revision,application:intended.application.id,applicationVersion:intended.application.version};
  if(!object(observed))fail('EPX_IDENTITY_UNVERIFIABLE','unavailable','BIND-003','Provider identity metadata is unavailable.',{dependency});
  for(const [key,value] of Object.entries(expected))if(observed[key]!==value)fail('EPX_PROVIDER_MISMATCH','unavailable','BIND-005','Observed provider identity/version does not match the intended dependency.',{dependency,expected,observed});
  if(!intended.executionPlatforms.includes(observed.executionPlatform))fail('EPX_PLATFORM_UNAVAILABLE','unavailable','BIND-005','Execution target platform is incompatible.',{dependency,expected:intended.executionPlatforms,observed:observed.executionPlatform});
  return observed;
}

export async function openBinding(pkg,configuration,reference,configDirectory) {
  const intended=pkg.requirements.bindings[reference];
  const local=configuration.bindings?.[reference];
  if(!intended)fail('EPX_BINDING_INVALID','invalid','BIND-001','Unknown logical binding.',{dependency:reference});
  if(!local)fail('EPX_PROVIDER_UNAVAILABLE','unavailable','BIND-005','Configure access to the intended provider.',{dependency:reference});
  if(!deepEqual(local.implementation,{id:intended.implementation.id,revision:intended.implementation.revision})||!deepEqual(local.application,intended.application))
    fail('EPX_BINDING_MISMATCH','unavailable','BIND-002','Local binding names another implementation or application.',{dependency:reference});
  if(intended.protocolVersion!==PROTOCOL)fail('EPX_PROTOCOL_UNSUPPORTED','unavailable','BIND-003','The declared MCP dependency revision is unavailable to this adapter.',{dependency:reference,expected:PROTOCOL,observed:intended.protocolVersion});
  const transport=structuredClone(local.transport);
  if(transport.kind==='stdio') {
    if(local.trust?.kind!=='launcher-sha256')fail('EPX_TRUST_UNVERIFIABLE','unavailable','BIND-002','Stdio fixture requires explicit launcher digest evidence.',{dependency:reference});
    const launcher=path.resolve(configDirectory,local.trust.path);
    let digest;try{digest=sha256(await readFile(launcher));}catch{fail('EPX_PROVIDER_UNAVAILABLE','unavailable','BIND-002','The intended launcher file is unavailable.',{dependency:reference});}
    if(digest!==local.trust.sha256||digest!==intended.implementation.artifactSha256)fail('EPX_LAUNCHER_MISMATCH','unavailable','BIND-002','Launcher bytes do not match intended and configured identity evidence.',{dependency:reference,expected:intended.implementation.artifactSha256,observed:digest});
    const invocation=[transport.command,...(transport.arguments??[])];
    if(!invocation.some(argument=>path.resolve(configDirectory,argument)===launcher))fail('EPX_TRUST_UNVERIFIABLE','unavailable','BIND-002','Verified launcher is absent from the explicit invocation.',{dependency:reference});
  } else if(transport.kind==='http') {
    if(local.trust?.kind!=='configured-service'||local.trust.subject!==intended.implementation.id||!local.trust.description)
      fail('EPX_TRUST_UNVERIFIABLE','unavailable','BIND-002','HTTP requires a configured intended service trust boundary.',{dependency:reference});
    let url;try{url=new URL(transport.url);}catch{fail('EPX_ENDPOINT_INVALID','unavailable','BIND-002','Configured HTTP endpoint is invalid.',{dependency:reference});}
    if(!['http:','https:'].includes(url.protocol)||url.username||url.password)fail('EPX_ENDPOINT_INVALID','unavailable','BIND-002','Endpoint must be HTTP(S) with credentials supplied separately.',{dependency:reference});
  }
  const client=new MCPClient(transport,{directory:configDirectory});
  try {
    const initialized=await client.initialize();
    const observed=providerMetadata(initialized,intended,reference);
    const tools=[];let cursor;const cursors=new Set();
    do{
      const listed=await client.request('tools/list',cursor?{cursor}:{});providerMetadata(listed,intended,reference);
      if(!Array.isArray(listed.tools))fail('EPX_TOOLS_INVALID','unavailable','BIND-004','Provider tools/list is malformed.',{dependency:reference});
      tools.push(...listed.tools);cursor=listed.nextCursor;
      if(cursor){if(cursors.has(cursor)||cursors.size>=100)fail('EPX_TOOLS_INVALID','unavailable','BIND-004','Provider tool pagination did not terminate.',{dependency:reference});cursors.add(cursor);}
    }while(cursor);
    const names=tools.map(t=>t.name);if(new Set(names).size!==names.length)fail('EPX_TOOLS_INVALID','unavailable','BIND-004','Provider advertised duplicate tool names.',{dependency:reference});
    for(const op of Object.values(pkg.requirements.operations).filter(o=>o.binding===reference)) {
      const pinned=pkg.contracts[op.toolContract];const discovered=tools.find(tool=>tool.name===op.tool);
      if(!discovered)fail('EPX_TOOL_UNAVAILABLE','unavailable','BIND-004','The exact intended tool is unavailable.',{dependency:reference,expected:op.tool});
      if(!deepEqual({name:discovered.name,inputSchema:discovered.inputSchema,outputSchema:discovered.outputSchema},{name:pinned.name,inputSchema:pinned.inputSchema,outputSchema:pinned.outputSchema}))
        fail('EPX_TOOL_CONTRACT_MISMATCH','unavailable','BIND-004','Discovered tool schema differs from its pinned descriptor.',{dependency:reference,expected:op.tool});
    }
    const resources=[];
    for(const [id,resource] of Object.entries(pkg.requirements.resources).filter(([,r])=>r.binding===reference)) {
      if(!object(initialized.capabilities?.resources))fail('EPX_RESOURCE_UNAVAILABLE','unavailable','RES-001','Provider does not advertise resources capability.',{dependency:id});
      let response;
      try{response=await client.request('resources/read',{uri:resource.uri});}catch(error){
        const cause=error.diagnostic?.cause;
        const code=cause?.code===-32002?'EPX_RESOURCE_ABSENT':cause?.code===-32003?'EPX_RESOURCE_DENIED':'EPX_RESOURCE_UNAVAILABLE';
        fail(code,'unavailable','RES-002','Required resource cannot be verified; configure access to the intended object.',{dependency:id,cause:cause??error.message});
      }
      const content=response?.contents?.find(c=>c.uri===resource.uri&&typeof c.text==='string');let identity;
      try{identity=strictJSON(content.text);}catch{fail('EPX_RESOURCE_UNVERIFIABLE','unavailable','RES-001','Resource did not supply its declared JSON identity evidence.',{dependency:id});}
      if(identity.id!==resource.uri||identity.provider!==intended.application.id)fail('EPX_RESOURCE_MISMATCH','unavailable','RES-001','Observed resource identity is not the intended object/provider.',{dependency:id,observed:identity});
      if(identity.revision!==resource.revision)fail('EPX_RESOURCE_REVISION_MISMATCH','unavailable','RES-001','Required resource revision is unavailable.',{dependency:id,expected:resource.revision,observed:identity.revision});
      resources.push({id,uri:identity.id,provider:identity.provider,revision:identity.revision});
    }
    return {client,intended,observed,resources,tools};
  }catch(error){await client.close();if(error instanceof Failure)throw error;fail('EPX_PROVIDER_UNAVAILABLE','unavailable','BIND-005','Provider preflight could not complete.',{dependency:reference,cause:error.message});}
}

File: src/epx/mcp.py
"""Independent MCP 2025-11-25 client for the declared stdio/HTTP-JSON subset."""

from __future__ import annotations

import os
from pathlib import Path
import queue
import subprocess
import threading
import urllib.error
import urllib.parse
import urllib.request

from . import VERSION
from . import jsonio
from .errors import Problem, invalid, unavailable

PROTOCOL = "2025-11-25"
META_KEY = "engineering-process-provider"


class RpcFailure(Exception):
    def __init__(self, error):
        super().__init__(str(error))
        self.error = error


class LostResponse(Exception):
    pass


class Client:
    def __init__(self, local, timeout=10):
        self.local = local
        self.transport = local["transport"]
        self.timeout = timeout
        self.next_id = 0
        self.session = None
        self.process = None
        self.responses = queue.Queue()
        self.last_call_sent = False
        if self.transport["kind"] == "stdio":
            try:
                self.process = subprocess.Popen(
                    [self.transport["command"], *self.transport["arguments"]],
                    stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.DEVNULL,
                )
            except OSError as exc:
                raise unavailable("PROVIDER_START", "BIND-002", "Cannot start configured MCP provider", cause=str(exc)) from exc
            def receive():
                try:
                    for line in self.process.stdout:
                        self.responses.put(line)
                finally:
                    self.responses.put(None)
            threading.Thread(target=receive, daemon=True).start()

    def close(self):
        if self.process:
            try:
                self.process.stdin.close()
                self.process.wait(timeout=1)
            except (OSError, subprocess.TimeoutExpired):
                self.process.kill()
                self.process.wait(timeout=2)

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self.close()

    def send(self, message, expect_response):
        payload = jsonio.encode(message).replace(b"\n", b"") + b"\n"
        # JSON string newlines remain escaped; remove only encoder formatting.
        if self.transport["kind"] == "stdio":
            try:
                self.process.stdin.write(payload)
                self.process.stdin.flush()
                if message.get("method") == "tools/call":
                    self.last_call_sent = True
            except OSError as exc:
                raise LostResponse("MCP connection closed while sending") from exc
            if not expect_response:
                return None
            while True:
                try:
                    line = self.responses.get(timeout=self.timeout)
                except queue.Empty as exc:
                    raise LostResponse("MCP response timed out") from exc
                if line is None:
                    raise LostResponse("MCP connection closed before its response")
                try:
                    response = jsonio.loads(line)
                except ValueError as exc:
                    raise RpcFailure({"message": "Invalid MCP JSON response", "detail": str(exc)}) from exc
                if isinstance(response, dict) and "method" in response and "id" not in response:
                    continue
                return response
        headers = {"Content-Type": "application/json", "Accept": "application/json, text/event-stream", "MCP-Protocol-Version": PROTOCOL}
        if self.session:
            headers["Mcp-Session-Id"] = self.session
        if self.transport.get("tokenEnv"):
            token = os.environ.get(self.transport["tokenEnv"])
            if not token:
                raise unavailable("CREDENTIAL_UNAVAILABLE", "BIND-002", "Receiving credential environment variable is absent")
            headers["Authorization"] = "Bearer " + token
        request = urllib.request.Request(self.transport["url"], data=payload, headers=headers, method="POST")
        try:
            if message.get("method") == "tools/call":
                self.last_call_sent = True
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                self.session = response.headers.get("Mcp-Session-Id", self.session)
                content = response.read()
                if not expect_response:
                    return None
                if response.headers.get_content_type() != "application/json":
                    raise RpcFailure({"message": "Only the declared MCP HTTP JSON-response subset is supported"})
                return jsonio.loads(content)
        except urllib.error.HTTPError as exc:
            body = exc.read()
            try:
                error = jsonio.loads(body)
            except ValueError:
                error = {"message": "HTTP failure", "status": exc.code}
            raise RpcFailure(error) from exc
        except (urllib.error.URLError, TimeoutError, ConnectionError, OSError) as exc:
            raise LostResponse("MCP HTTP connection failed before a verified response") from exc
        except ValueError as exc:
            raise RpcFailure({"message": "Invalid MCP JSON response", "detail": str(exc)}) from exc

    def request(self, method, params=None):
        self.next_id += 1
        message = {"jsonrpc": "2.0", "id": self.next_id, "method": method}
        if params is not None:
            message["params"] = params
        result = self.send(message, True)
        if not isinstance(result, dict) or result.get("jsonrpc") != "2.0" or result.get("id") != self.next_id:
            raise RpcFailure({"message": "Invalid JSON-RPC response envelope or request identity"})
        if "error" in result:
            raise RpcFailure(result["error"])
        if "result" not in result:
            raise RpcFailure({"message": "Missing JSON-RPC result"})
        return result["result"]

    def initialize(self):
        response = self.request("initialize", {
            "protocolVersion": PROTOCOL, "capabilities": {},
            "clientInfo": {"name": "epx-python", "version": VERSION},
        })
        if response.get("protocolVersion") != PROTOCOL:
            raise unavailable("PROTOCOL_MISMATCH", "BIND-003", "MCP protocol did not match", expected=PROTOCOL, observed=response.get("protocolVersion"))
        if "tools" not in response.get("capabilities", {}):
            raise unavailable("CAPABILITY_UNAVAILABLE", "BIND-003", "MCP provider does not declare tools capability")
        self.send({"jsonrpc": "2.0", "method": "notifications/initialized"}, False)
        return response


def configuration(path: Path):
    value = jsonio.read(path)
    jsonio.validate(value, jsonio.load_schema("bindings"), code="BINDINGS_INVALID", requirement="BIND-002")
    return value


def verify_local(required, local, ref):
    for key, fields in (("implementation", ("id", "revision")), ("application", ("id", "version"))):
        for field in fields:
            if local[key][field] != required[key][field]:
                raise unavailable("PROVIDER_IDENTITY_MISMATCH", "BIND-001", "Configured provider differs from author requirement", dependency=ref, expected=required[key], observed=local[key])
    if required["protocolVersion"] != PROTOCOL:
        raise unavailable("PROTOCOL_UNSUPPORTED", "BIND-003", "Adapter does not implement the required MCP revision", dependency=ref, expected=required["protocolVersion"], observed=PROTOCOL)
    trust, transport = local["trust"], local["transport"]
    if transport["kind"] == "stdio":
        if trust["kind"] != "launcher-sha256" or not required["implementation"].get("artifactSha256"):
            raise unavailable("IDENTITY_EVIDENCE_MISSING", "BIND-002", "Stdio requires author-pinned launcher evidence", dependency=ref)
        path = Path(trust["path"]).resolve()
        try:
            actual = jsonio.digest(path.read_bytes())
        except OSError as exc:
            raise unavailable("LAUNCHER_UNAVAILABLE", "BIND-002", "Configured launcher cannot be read", dependency=ref) from exc
        arguments = [Path(arg).resolve() for arg in transport["arguments"] if not arg.startswith("-")]
        if actual != trust["sha256"] or actual != required["implementation"]["artifactSha256"] or path not in arguments:
            raise unavailable("LAUNCHER_IDENTITY_MISMATCH", "BIND-002", "Launcher invocation/digest does not establish intended implementation", dependency=ref, observed=actual, expected=required["implementation"]["artifactSha256"])
    else:
        url = urllib.parse.urlparse(transport["url"])
        if url.username or url.password:
            raise unavailable("CONNECTION_CREDENTIALS", "BIND-002", "Use receiving environment credentials, not URL userinfo", dependency=ref)
        if trust["kind"] != "configured-service" or trust["subject"] != required["implementation"]["id"]:
            raise unavailable("SERVICE_TRUST_MISSING", "BIND-002", "Remote service requires explicit intended-provider trust", dependency=ref)


def verify_metadata(required, message, ref):
    observed = message.get("_meta", {}).get(META_KEY)
    expected = {
        "implementation": required["implementation"]["id"],
        "revision": required["implementation"]["revision"],
        "application": required["application"]["id"],
        "applicationVersion": required["application"]["version"],
    }
    if not isinstance(observed, dict) or any(observed.get(key) != value for key, value in expected.items()):
        raise unavailable("PROVIDER_METADATA_MISMATCH", "BIND-003", "Observed MCP/application identity differs or is unverifiable", dependency=ref, expected=expected, observed=observed)
    if observed.get("executionPlatform") not in required["executionPlatforms"]:
        raise unavailable("PLATFORM_UNAVAILABLE", "BIND-005", "Intended execution platform is unsupported", dependency=ref, expected=required["executionPlatforms"], observed=observed.get("executionPlatform"))
    return observed


def verify_session(package, config, ref, client):
    required = package.requirements["bindings"][ref]
    observed = verify_metadata(required, client.initialize(), ref)
    tools = []
    cursor = None
    seen = set()
    while True:
        listing = client.request("tools/list", {"cursor": cursor} if cursor else {})
        verify_metadata(required, listing, ref)
        tools.extend(listing.get("tools", []))
        cursor = listing.get("nextCursor")
        if not cursor:
            break
        if cursor in seen:
            raise unavailable("TOOL_LIST_INVALID", "BIND-004", "Provider repeats discovery cursor", dependency=ref)
        seen.add(cursor)
    names = [tool.get("name") for tool in tools]
    if len(set(names)) != len(names):
        raise unavailable("TOOL_LIST_INVALID", "BIND-004", "Provider lists duplicate tool identities", dependency=ref)
    by_name = {tool["name"]: tool for tool in tools}
    for node, operation in package.requirements["operations"].items():
        if operation["binding"] != ref:
            continue
        actual = by_name.get(operation["tool"])
        expected = jsonio.read(package.payload(operation["toolContract"]))
        if actual is None:
            raise unavailable("TOOL_UNAVAILABLE", "BIND-004", "Required tool is not available on intended provider", node=node, dependency=ref, expected=operation["tool"])
        for field in ("name", "inputSchema", "outputSchema"):
            if not jsonio.equal(expected.get(field), actual.get(field)):
                raise unavailable("TOOL_CONTRACT_MISMATCH", "BIND-004", "Discovered tool contract changed", node=node, dependency=ref, field=field)
    resources = []
    for resource_id, requirement in package.requirements["resources"].items():
        if requirement["binding"] != ref:
            continue
        try:
            result = client.request("resources/read", {"uri": requirement["uri"]})
            contents = result["contents"]
            matching = [content for content in contents if content.get("uri") == requirement["uri"] and "text" in content]
            if len(matching) != 1:
                raise ValueError("Exactly one matching JSON resource representation is required")
            identity = jsonio.loads(matching[0]["text"])
        except (RpcFailure, KeyError, TypeError, ValueError) as exc:
            cause = exc.error if isinstance(exc, RpcFailure) else str(exc)
            raise unavailable("RESOURCE_UNAVAILABLE", "RES-002", "Required resource cannot be verified", dependency=resource_id, expected=requirement, cause=cause) from exc
        expected = {"id": requirement["uri"], "provider": required["application"]["id"], "revision": requirement["revision"]}
        if not isinstance(identity, dict) or any(identity.get(key) != value for key, value in expected.items()):
            raise unavailable("RESOURCE_REVISION_MISMATCH", "RES-001", "Observed resource identity/revision differs", dependency=resource_id, expected=expected, observed=identity)
        resources.append({"requirement": resource_id, **expected})
    return observed, resources


def preflight(package, config):
    observations = []
    for ref, required in package.requirements["bindings"].items():
        local = config["bindings"].get(ref)
        if local is None:
            raise unavailable("PROVIDER_UNAVAILABLE", "BIND-005", "No local connection is configured for the intended binding", dependency=ref, expected=required)
        verify_local(required, local, ref)
        try:
            with Client(local) as client:
                observed, resources = verify_session(package, config, ref, client)
                observations.append({"binding": ref, "provider": observed, "resources": resources})
        except (RpcFailure, LostResponse) as exc:
            cause = exc.error if isinstance(exc, RpcFailure) else str(exc)
            raise unavailable("PROVIDER_UNAVAILABLE", "BIND-005", "Provider preflight failed", dependency=ref, cause=cause) from exc
    return {"status": "ready", "bindings": observations, "diagnostics": [], "engineeringCalls": 0}
